pkg/gofr/datasource/pubsub/mqtt/mqtt.go

```diff
@@ -143,23 +168,40 @@ func getClientID(clientID string) string {
 }
 
 func (m *MQTT) Subscribe(ctx context.Context, topic string) (*pubsub.Message, error) {
-	ctx, span := otel.GetTracerProvider().Tracer("gofr").Start(ctx, "mqtt-subscribe")
-	defer span.End()
-
-	m.metrics.IncrementCounter(ctx, "app_pubsub_subscribe_total_count", "topic", topic)
-
-	var messg = pubsub.NewMessage(ctx)
-
 	m.mu.Lock()
 	// get the message channel for the given topic
-	msgChan, ok := m.msgChanMap[topic]
+	subs, ok := m.subscriptions[topic]
 	if !ok {
-		msgChan = make(chan *pubsub.Message, messageBuffer)
-		m.msgChanMap[topic] = msgChan
+		subs.msgs = make(chan *pubsub.Message, messageBuffer)
+		subs.handler = m.createMqttHandler(ctx, topic, subs.msgs)
+		token := m.Client.Subscribe(topic, m.config.QoS, subs.handler)
+
+		if token.Wait() && token.Error() != nil {
+			m.logger.Errorf("error getting a message from MQTT, error: %v", token.Error())
+			return nil, token.Error()
+		}
+
+		m.subscriptions[topic] = subs
 	}
 	m.mu.Unlock()
 
-	handler := func(_ mqtt.Client, msg mqtt.Message) {
+	// blocks if there are no messages in the channel
+	msg := <-subs.msgs
+	m.metrics.IncrementCounter(msg.Context(), "app_pubsub_subscribe_success_count", "topic", msg.Topic)
+
+	return msg, nil
+}
+
+func (m *MQTT) createMqttHandler(_ context.Context, topic string, msgs chan *pubsub.Message) mqtt.MessageHandler {
+	return func(_ mqtt.Client, msg mqtt.Message) {
+		ctx := context.Background()
+		ctx, span := otel.GetTracerProvider().Tracer("gofr").Start(ctx, "mqtt-subscribe")
+
```

Comment:
tracing should be done when a message comes and not when we are registering a handler